client/me/developer/features/test/use-handle-click-link.js

```diff
@@ -0,0 +1,44 @@
+/**
+ * @jest-environment jsdom
+ */
+import { act, renderHook } from '@testing-library/react';
+import { useHandleClickLink } from '../use-handle-click-link';
+import { useRecordTracksEventWithUserIsDevAccount } from '../use-record-tracks-event-with-user-is-dev-account';
+
+jest.mock( '../use-record-tracks-event-with-user-is-dev-account', () => ( {
+	useRecordTracksEventWithUserIsDevAccount: jest.fn(),
+} ) );
+
+describe( 'useHandleClickLink', () => {
+	let mockRecordTracksEventWithUserIsDevAccount;
+
+	const tracksHandle = 'calypso_me_developer_learn_more';
+
+	beforeAll( () => {
+		mockRecordTracksEventWithUserIsDevAccount = jest.fn();
+		useRecordTracksEventWithUserIsDevAccount.mockReturnValue(
+			mockRecordTracksEventWithUserIsDevAccount
+		);
+	} );
+
+	describe( 'ensure track events are recorded without the /support/ prefix', () => {
+		test.each( [
+			[ 'https://wordpress.com/support/feature', 'feature' ],
+			[ 'https://wordpress.com/support/category/feature', 'category/feature' ],
+			[ 'https://wordpress.com/support/feature-with-hyphens', 'feature-with-hyphens' ],
+			[ 'https://wordpress.com/support/supported/feature', 'supported/feature' ],
+		] )(
```

Comment:
Thanks @Aurorum for adding unit tests to validate analytics values! 

>The idea is that it would prevent someone adding a link with the wrong structure that would break the analytics in future.

I'm not sure if we should restrict the structure of the link. We should allow the flexibility to supply any link. If the current analytics logic doesn't handle that link, then we would adjust the analytics logic accordingly when we cross the bridge, while maintaining the existing values that have been used to track events.

At the end of the day, what we need to ensure is that existing tracking values remain in tact when changes are made. So I would go as simple as listing out all the links we have and their expected tracking value, instead of trying to check for the removal of `/support` prefix. 

```suggestion
	describe( 'ensure track events are recorded with the expected feature name', () => {
		test.each( [
			[ 'https://wordpress.com/support/connect-to-ssh-on-wordpress-com', 'connect-to-ssh-on-wordpress-com', ],
			[ 'https://wordpress.com/support/how-to-create-a-staging-site/', 'how-to-create-a-staging-site', ],
		],
			[ 'https://wordpress.com/support/code', 'code', ],
			[ 'https://wordpress.com/support/domains/https-ssl', 'domains/https-ssl', ],
			[ 'https://wordpress.com/support/help-support-options', 'help-support-options', ],
			[ 'https://wordpress.com/support/restore', 'restore', ],
			[ 'https://wordpress.com/support/malware-and-site-security', 'malware-and-site-security', ],
			[ 'https://wordpress.com/support/site-monitoring', 'site-monitoring', ],
		] )(
```

Reply:
Hi @wongasy, nice to meet you 👋 

The two unit tests I've proposed are for `useHandleClickLink`. This hook, by design, has no knowledge of the specific links on the `/me/developer` page. It does however contain explicit handling for the `/support/` prefix and trailing `/`, and so it's this logic that the unit tests cover. ie.

The test "ensure track events are recorded without the /support/ prefix" covers these lines:

https://github.com/Automattic/wp-calypso/blob/582dcb267934545211826b17e1267d3b7f737402/client/me/developer/features/use-handle-click-link.tsx#L9-L15

While the test "ensure track events are recorded without a trailing /" covers this line:

https://github.com/Automattic/wp-calypso/blob/582dcb267934545211826b17e1267d3b7f737402/client/me/developer/features/use-handle-click-link.tsx#L16

To verify the tracking values for the _actual_ links on the `/me/developer` page, I would suggest creating a test for the `DeveloperFeatures` component. This component has knowledge of the specific links that we want to validate the tracking values for (via it's internal call to `useFeaturesList()`). The test would roughly take the form:

- Render the `DeveloperFeatures` component (using the same approach to mock `useRecordTracksEventWithUserIsDevAccount`)
- Click each link eg. `https://wordpress.com/support/connect-to-ssh-on-wordpress-com`
- Verify the tracking values

Reply:
Thanks @oswian for the explainer. I see your point that, it makes better sense to test the actual links in relation to the component that defines the links, instead of the helper function which doesn't have knowledge of the links.